api/web3server_utils.go

```diff
@@ -0,0 +1,645 @@
+	from, err = svr.parseBlockNumber(fromStr)
+	if err != nil {
+		return
+	}
+	to, err = svr.parseBlockNumber(toStr)
+	if err != nil {
+		return
+	}
+	tipHeight := svr.coreService.bc.TipHeight()
+	if from > tipHeight {
+		err = status.Error(codes.InvalidArgument, "start block > tip height")
+		return
+	}
+	if to > tipHeight {
+		to = tipHeight
+	}
+	return
+}
+
+func (svr *Web3Server) isContractAddr(addr string) (bool, error) {
+	if addr == "" {
+		return true, nil
+	}
+	accountMeta, _, err := svr.coreService.Account(addr)
+	if err != nil {
+		return false, err
+	}
+	return accountMeta.IsContract, nil
+}
+
+func (svr *Web3Server) getLogsWithFilter(from uint64, to uint64, addrs []string, topics [][]string) ([]logsObject, error) {
+	// construct filter topics and addresses
+	var filter iotexapi.LogsFilter
+	for _, ethAddr := range addrs {
+		ioAddr, err := ethAddrToIoAddr(ethAddr)
+		if err != nil {
+			return nil, err
+		}
+		filter.Address = append(filter.Address, ioAddr)
+	}
+	for _, tp := range topics {
+		var topic [][]byte
+		for _, str := range tp {
+			b, err := hex.DecodeString(str)
+			if err != nil {
+				return nil, err
+			}
+			topic = append(topic, b)
+		}
+		filter.Topics = append(filter.Topics, &iotexapi.Topics{
+			Topic: topic,
+		})
+	}
+	logs, err := svr.coreService.getLogsInRange(logfilter.NewLogFilter(&filter, nil, nil), from, to, 1000)
+	if err != nil {
+		return nil, err
+	}
+
+	// parse log results
+	var ret []logsObject
+	for _, l := range logs {
+		var topics []string
+		for _, val := range l.Topics {
+			topics = append(topics, byteToHex(val))
+		}
+		contractAddr, err := ioAddrToEthAddr(l.ContractAddress)
+		if err != nil {
+			return nil, err
+		}
+		ret = append(ret, logsObject{
+			BlockHash:       byteToHex(l.BlkHash),
+			TransactionHash: byteToHex(l.ActHash),
+			LogIndex:        uint64ToHex(uint64(l.Index)),
+			BlockNumber:     uint64ToHex(l.BlkHeight),
+			// TransactionIndex bug will be fixed in the next
+			TransactionIndex: "0x1",
+			Address:          contractAddr,
+			Data:             byteToHex(l.Data),
+			Topics:           topics,
+		})
+	}
+	return ret, nil
+}
+
+func byteToHex(b []byte) string {
+	return "0x" + hex.EncodeToString(b)
+}
+
+func parseLogRequest(in gjson.Result) (*filterObject, error) {
+	var logReq filterObject
+	if len(in.Array()) > 0 {
+		req := in.Array()[0]
+		logReq.FromBlock = req.Get("fromBlock").String()
+		logReq.ToBlock = req.Get("toBlock").String()
+		for _, addr := range req.Get("address").Array() {
+			logReq.Address = append(logReq.Address, addr.String())
+		}
+		for _, topics := range req.Get("topics").Array() {
+			if topics.IsArray() {
+				var topicArr []string
+				for _, topic := range topics.Array() {
+					topicArr = append(topicArr, removeHexPrefix(topic.String()))
+				}
+				logReq.Topics = append(logReq.Topics, topicArr)
+			} else {
+				logReq.Topics = append(logReq.Topics, []string{removeHexPrefix(topics.String())})
```

Comment:
so topics does not need 0x prefix, can we make sure that's correct by web3 spec?